patches/server/0445-Rewrite-custom-payload-handler.patch

```diff
@@ -0,0 +1,180 @@
++    public static final ResourceLocation UNREGISTER = new ResourceLocation("unregister");
++
++    private final String channels;
++
++    private ChannelPayload(final FriendlyByteBuf buf) {
++        this.channels = buf.toString(StandardCharsets.UTF_8);
++        buf.readerIndex(buf.readerIndex() + channels.length());
++    }
++
++    @Override
++    public final void write(final FriendlyByteBuf buf) {
++        buf.writeBytes(channels.getBytes(StandardCharsets.UTF_8));
++    }
++
++    public final String[] channels() {
++        return channels.split("\0");
++    }
++
++    /**
++     * Channel register payload. (minecraft:register)
++     */
++    public static final class Register extends ChannelPayload {
++
++        public Register(final FriendlyByteBuf buf) {
++            super(buf);
++        }
++
++        @Override
++        public @NotNull ResourceLocation id() {
++            return REGISTER;
++        }
++    }
++
++    /**
++     * Channel unregister payload. (minecraft:unregister)
++     */
++    public static final class Unregister extends ChannelPayload {
++
++        public Unregister(final FriendlyByteBuf buf) {
++            super(buf);
++        }
++
++        @Override
++        public @NotNull ResourceLocation id() {
++            return UNREGISTER;
++        }
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/network/payload/ReadableMessage.java b/src/main/java/io/papermc/paper/network/payload/ReadableMessage.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..df2d63ab6c557755ed740c1b57e1a71af9db4ba2
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/network/payload/ReadableMessage.java
+@@ -0,0 +1,8 @@
++package io.papermc.paper.network.payload;
++
++import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
++
++public interface ReadableMessage extends CustomPacketPayload {
++
++    byte[] message();
++}
+diff --git a/src/main/java/net/minecraft/network/protocol/common/ServerboundCustomPayloadPacket.java b/src/main/java/net/minecraft/network/protocol/common/ServerboundCustomPayloadPacket.java
+index af86f752c33a2990405fea058b7c41c437ba9d46..8f5ab7d6cceb3558960eeea1b90a75d227ba367a 100644
+--- a/src/main/java/net/minecraft/network/protocol/common/ServerboundCustomPayloadPacket.java
++++ b/src/main/java/net/minecraft/network/protocol/common/ServerboundCustomPayloadPacket.java
+@@ -12,7 +12,13 @@ import net.minecraft.resources.ResourceLocation;
+ public record ServerboundCustomPayloadPacket(CustomPacketPayload payload) implements Packet<ServerCommonPacketListener> {
+ 
+     private static final int MAX_PAYLOAD_SIZE = 32767;
+-    private static final Map<ResourceLocation, FriendlyByteBuf.Reader<? extends CustomPacketPayload>> KNOWN_TYPES = ImmutableMap.<ResourceLocation, FriendlyByteBuf.Reader<? extends CustomPacketPayload>>builder().build(); // CraftBukkit - no special handling
++    // Paper start - Rewrite custom payload handler
```

Comment:
I generally want some comments on what best to do here, 32kb of data is pretty small by default and so I did maybe envision that we'd add way to allow plugins to register their own channels AND override this limit, having this this thing before, it's a PITA as the network I was doing stuff for really didn't wanna pick up additional infra, for ex